zipline/gens/brokers/alpaca_broker.py

```diff
@@ -0,0 +1,266 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import alpaca_trade_api as tradeapi
+from zipline.gens.brokers.broker import Broker
+import zipline.protocol as zp
+from zipline.finance.order import (Order as ZPOrder,
+                                   ORDER_STATUS as ZP_ORDER_STATUS)
+from zipline.finance.execution import (MarketOrder,
+                                       LimitOrder,
+                                       StopOrder,
+                                       StopLimitOrder)
+from zipline.api import symbol as symbol_lookup
+from zipline.errors import SymbolNotFound
+import pandas as pd
+import uuid
+
+from logbook import Logger
+import sys
+
+if sys.version_info > (3,):
+    long = int
+
+log = Logger('Alpaca Broker')
+NY = 'America/New_York'
+
+
+class ALPACABroker(Broker):
+    '''
+    Broker class for Alpaca.
+    The uri parameter is not used. Instead, the API key must be
+    set via environment variables (APCA_API_KEY_ID and APCA_API_SECRET_KEY).
+    Orders are identified by the UUID (v4) generated here and
+    associated in the broker side using client_order_id attribute.
+    Currently this class makes use of REST API only, but websocket
+    streaming can possibly used too.
+    '''
+
+    def __init__(self, uri):
+        self._api = tradeapi.REST()
+
+    def subscribe_to_market_data(self, asset):
+        '''Do nothing to comply the interface'''
+        pass
+
+    def subscribed_assets(self):
+        '''Do nothing to comply the interface'''
+        pass
+
+    def _map_bars(self, symbols, timeframe, **kwargs):
+        bars_list = self._api.list_bars(symbols, timeframe, **kwargs)
+        results = {}
+        for abars in bars_list:
+            results[abars.symbol] = abars.bars
+        return results
+
+    @property
+    def positions(self):
+        z_positions = zp.Positions()
+        positions = self._api.list_positions()
+        position_map = {}
+        symbols = []
+        for pos in positions:
+            symbol = pos.symbol
+            try:
+                z_position = zp.Position(symbol_lookup(symbol))
+            except SymbolNotFound:
+                continue
+            z_position.amount = pos.qty
+            z_position.cost_basis = float(pos.cost_basis)
+            z_position.last_sale_price = None
+            z_position.last_sale_date = None
+            z_positions[symbol_lookup(symbol)] = z_position
+            symbols.append(symbol)
+            position_map[symbol] = z_position
+
+        quotes = self._api.list_quotes(symbols)
+        for quote in quotes:
+            price = quote.last
+            dt = quote.last_timestamp
+            z_position = position_map[quote.symbol]
+            z_position.last_sale_price = float(price)
+            z_position.last_sale_date = dt
+        return z_positions
```

Comment:
here you loop twice through the positions/ quotes, can't we do this in one go? with a small extension in self._api.list_positions() we could do it in one go... or do I see this wrong?

Reply:
Well, in order to get quotes efficiently, I need to gather symbols as a list first, and considering possibility some symbol may not make it to the list, list of quotes and positions may not be 1:1.  In any case, loop twice is just O(N) as the matching is hash lookup by symbol string (O(1))